modules/core/src/main/java/org/apache/ignite/spi/encryption/keystore/KeystoreEncryptionSpi.java

```diff
@@ -0,0 +1,502 @@
+            SecretKeySpec keySpec = new SecretKeySpec(((KeystoreEncryptionKey)key).key().getEncoded(), CIPHER_ALGO);
+
+            Cipher cipher = aesWithPadding.get();
+
+            cipher.init(DECRYPT_MODE, keySpec, new IvParameterSpec(data, 0, cipher.getBlockSize()));
+
+            return cipher.doFinal(data, cipher.getBlockSize(), data.length - cipher.getBlockSize());
+        }
+        catch (InvalidAlgorithmParameterException | InvalidKeyException | IllegalBlockSizeException |
+            BadPaddingException e) {
+            throw new IgniteSpiException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void decryptNoPadding(ByteBuffer data, Serializable key, ByteBuffer res) {
+        assert key instanceof KeystoreEncryptionKey;
+
+        ensureStarted();
+
+        try {
+            SecretKeySpec keySpec = new SecretKeySpec(((KeystoreEncryptionKey)key).key().getEncoded(), CIPHER_ALGO);
+
+            Cipher cipher = aesWithoutPadding.get();
+
+            byte[] iv = new byte[cipher.getBlockSize()];
+
+            data.get(iv);
+
+            cipher.init(DECRYPT_MODE, keySpec, new IvParameterSpec(iv));
+
+            cipher.doFinal(data, res);
+        }
+        catch (InvalidAlgorithmParameterException | InvalidKeyException | IllegalBlockSizeException |
+            ShortBufferException | BadPaddingException e) {
+            throw new IgniteSpiException(e);
+        }
+    }
+
+    /**
+     * @param data Plain data.
+     * @param cipher Cipher.
+     * @param key Encryption key.
+     */
+    private void doEncryption(ByteBuffer data, Cipher cipher, Serializable key, ByteBuffer res) {
+        assert key instanceof KeystoreEncryptionKey;
+        //assert start >= 0 && length + start <= data.length;
+
+        ensureStarted();
+
+        try {
+            SecretKeySpec keySpec = new SecretKeySpec(((KeystoreEncryptionKey)key).key().getEncoded(), CIPHER_ALGO);
+
+            byte[] iv = initVector(cipher);
+
+            res.put(iv);
+
+            cipher.init(ENCRYPT_MODE, keySpec, new IvParameterSpec(iv));
+
+            cipher.doFinal(data, res);
+        }
+        catch (ShortBufferException | InvalidAlgorithmParameterException | InvalidKeyException |
+            IllegalBlockSizeException | BadPaddingException e) {
+            throw new IgniteSpiException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public byte[] encryptKey(Serializable key) {
+        assert key instanceof KeystoreEncryptionKey;
+
+        byte[] serKey = U.toBytes(key);
+
+        byte[] res = new byte[encryptedSize(serKey.length)];
+
+        encrypt(ByteBuffer.wrap(serKey), masterKey, ByteBuffer.wrap(res));
+
+        return res;
+    }
+
+    /** {@inheritDoc} */
+    @Override public KeystoreEncryptionKey decryptKey(byte[] data) {
+        byte[] serKey = decrypt(data, masterKey);
+
+        KeystoreEncryptionKey key = U.fromBytes(serKey);
+
+        byte[] digest = makeDigest(key.key().getEncoded());
```

Comment:
Please note that this digest is not a digest of a key clear value, but some java defined serialized representation of the key. Probably ASN-1 encoded key or something like that.

Reply:
Thanks.